Twister: restructure and class split

nashif opened the pull request:
This is still work in progress, splitting twisterlib.py into smaller classes and move code from main twister script into respective classes.
Additionaly, moving lots of code condensed in testplan to different classes where things should belong, i.e. reporting, runner code etc.


- twister: optimize file scanning
- twister: add environment class
- twister: move cmake cache class
- twister: move testsuite, testcase classes out
- twister: split report classes out
- twister: split hardwaremap class out
- twister: move handlers class out
- twister: import cleanup
- twister: move platform class out
- twister: cleanup imports
- twister: move coverage classes out
- twister: move test instance class out
- twister: minor cleanups in main script
- twister: fix rebase
- twister: move runner classes out
- twister: isolate testplan class
- twister: cleanup option handling with testplan class
- twister: move execution code out of testplan
- twister: move runner code into runner class
- twister: more cleanup to option passing
- twister: rework testplan
- twister: testplan: move calls to json loading/filter into class
- twister: move cleanup and classification of code
- twister: move device testing code
- twister: move size report
- twister: move more code from main script
- twister: move harness class to class folder
- twister: simplify handler options
- twister: make ninja the default
- twister: fix subset processing
- twister: fix returns
- twister: testplan: get options directly

nashif committed:
twister: optimize file scanning

We have been scanning code for tests for every scenario defined in the
yaml file. This needs to only be parsed once and not more. We are
dealing with the same files for each scenario.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: add environment class

Add environment class to discover toolchain, zephyr version and other
items.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: move cmake cache class

Move cmake cache classes into own file.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: move testsuite, testcase classes out

Move testsuite/testcase and supporting classes into own files.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: split report classes out

Move report classes out into own file and split out of testplan class.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: split hardwaremap class out

Split hardware map class into own file.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: move handlers class out

Split handlers into own file.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: import cleanup

python import cleanup and simplification.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: move platform class out

Move both platform and config parser into own file.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: cleanup imports and move size classes out

More python import optimizations and cleanup and move size_calc class
out into own file.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: move coverage classes out

Move coverage related classes out into own file.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: move test instance class out

Move test instance class into own file.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: minor cleanups in main script

Move code around in the main script and cleanup some of the options.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: move runner classes out

Move runners into own file.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: isolate testplan class

Move testplan class into own file and rename the original twisterlib.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: move execution code out of testplan

Move all code related to execution into runner class.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: move runner code into runner class

Move all code related to test execution into runner class.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: more cleanup to option passing

Use global options rather than passing them individually.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: rework testplan and main twister script

Move all testplan related code to the testplan class and cleanup main
twister script.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: testplan: move calls to json loading/filter into class

Move all code related to instance filtering and loading into the class
instead of having it in the main script. Simplify option passing.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: move cleanup and classification of code

Move code out of the main script and use global options.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: move device testing code

Move device handling code into hardware map class.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: move size report into size calc class

Move from main script into size_calc class.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: move more code from main script

Move code from twister script into respective classes.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: move harness class to class folder

harness class moved along size other classes.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: simplify handler options

Depend on global options instead of passing them around for each class.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: make ninja the default

Using ninja is up to 50% faster, make it the default and add an option
to build with Make.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: fix subset processing

Process subsets was done in the wrong step, so make it apply and
generarte subsets correctly.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: fix main script returns and error handling

Propagate errors to the main routine and do not exit early in case of
errors.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: testplan: get options directly

Get options from passed environment class instead of using command line
options to the init function.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: fix twister testsuite

Make testsuite work again.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: move file scanning out of class

This is all static functions that can maintained outside of the class
and called only in one place, so move this to the testsuite file.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: adapt and fix testsuite tests

Re-add testsuite tests, include file scanning.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: fixed various pylint issues

Fix various pylint issues.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

chen-png commented on scripts/pylib/twister/twisterlib/reports.py:
```diff
@@ -0,0 +1,499 @@
+        logger.info(f"Writing JSON report {filename}")
+        report = {}
+        report["environment"] = {"os": os.name,
+                                 "zephyr_version": version,
```
self.env.version

nashif replied:
this is coming as an argument which is set to self.env.version already.

nashif committed:
twister: fixups to various files

Various fixups, will be folded in correct commits eventually.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

yperess reviewed (changes requested):
Wow, I learned a lot about twister today :)

PerMac commented on scripts/pylib/twister/twisterlib.py:
```diff
@@ -3346,127 +3074,398 @@ def add_configurations(self):
+        registered_suite_run_regex = re.compile(
+            br"^\s*ztest_run_registered_test_suites\("
+            br"(\*+|&)?(?P<state_identifier>[a-zA-Z0-9_]+)\)",
+            re.MULTILINE)
```
@nashif I am lost now. 
In #46052 you wrote
> This PR should work on generating the test plan and not as a post step to verify the test plan and set status. We could make this work on-demand by adding an option to the yaml file to generate list of test cases only on select test suites while we continue to parse things as before on everything else.

My understanding was that we will stay with mainly relying on parsing c files with twister to get a testplan fast by default, and use refine method e.g. based on .elf parsing if the "dumb" parser fails. 
The problem with parsing .elf is that you need this .elf, hence a successful build for that, but we want to know what will be tested (having test plan) before we start testing (building).

PerMac replied:
thanks for clarification. I have to admit I still don't understand, since for me the other way round seems more logical. Or maybe when we talk about parser we think about two different things. When I am writing parser, by default I mean a parser in twister, which is prone to final linker results pitfalls. IMO, since this is fast and doesn't require builds to finish, it can be used as the default option for getting test cases in the test plan. However, for special cases, where the tests in the binary doesn't correspond to the test plan, we can tell twister to align the plan based on the content of the binary (or some other method, e.g. explicitly adding testcases to yaml). Such cases could be exposed in advance thanks to #46052 and #46278. 
But anyway, I think we have the same goal in mind, i.e. reliable test plan with test cases. Maybe during a testing wg call I'll be able to better understand how we can get there.

yperess replied:
I would really love to understand this better. I'm not sure I'm seeing a benefit of having the test plan before the binary is built. Aaron will be attending the next testing working group meeting and I've asked to make sure that our testing team is aligned on this and we understand fully what the goal/benefit is.

nashif replied:
@yperess I am on the fence with this one. This was originally brought up here https://github.com/zephyrproject-rtos/zephyr/issues/28864 long time ago, but as we are all aware, it was and still  in most cases a testsuite configuration/setup problem rather than an issue with twister.

The only issue I see with this is when you have tests that are not building for whatever reason, in this case we will not be able to extract testcases. 

I do prefer where all testsuites/testcases are extracted in a consistent and complete way rather than depend on parsing like we do now. So if we figure out how to deal with such cases, I think we should be good.

nashif committed:
twister: rework testsuite parsing

Move testsuite setup and parsing to the testsuite class. Simplify
reading data from yaml.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
twister: adapt twister testsuite

Adapt with new way for parsing testsuites.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

enjiamai reviewed (approved):
Really nice restructuring and did not see issues on all our HW boards.
